EIPS/eip-5988.md

```diff
@@ -0,0 +1,159 @@
+- `alpha`: the power of S-box.
+- `input_rate`: the size of input.
+- `t`: the size of the state.
+- `full_round`: the number of full rounds. Denoted as `R_F` in the Poseidon paper.
+- `partial_round`: the number of partial rounds. Denoted as `R_P` in the Poseidon paper.
+
+The input to the precompile MUST be encoded as follows:
+
+- `p`: 32 bytes
+- `security_level`: 2 bytes
+- `alpha`: 1 byte
+- `input_rate`: 2 bytes
+- `t`: 1 byte
+- `full_round`: 1 byte
+- `partial_round`: 1 byte
+- `input`: `input_rate` \* 32 bytes
+
+```
+[32 bytes for p][2 bytes for security_level][1 byte for alpha][2 bytes for input_rate][1 byte for t][1 byte for full_round][1 byte for partial_round][input_rate * 32 bytes for input]
+```
+
+The precompile should compute the hash function as [specified in the Poseidon paper](../assets/eip-5988/papers/poseidon_paper.pdf) and return hash output.
+
+### Example Usage in Solidity
+
+The precompile can be wrapped easily in Solidity to provide a more development-friendly interface to `poseidon_hash` function.
+
+```solidity
+// TODO: Add solidity example
+```
+
+### Gas Costs
+
+```
+TODO: Fill gas costs section and add benchmarks
+```
+
+## Rationale
+
+TODO: Add rationale
+
+## Backwards Compatibility
+
+There is very little risk of breaking backwards-compatibility with this EIP, the sole issue being if someone were to build a contract relying on the address at `0xPOSEIDON_PRECOMPILE_ADDRESS` being empty. The likelihood of this is low, and should specific instances arise, the address could be chosen to be any arbitrary value with negligible risk of collision.
+
+## Test Cases
+
+The Poseidon reference implementation contains test vectors that can be used to test the precompile.
+Those tests are available [here](../assets/eip-5988/test/poseidon/test_vectors.txt).
+
+## Reference Implementation
+
+TODO: Add initial Geth implementation
+
+## Security Considerations
+
+Quoting Vitalik Buterin from `Arithmetic hash based alternatives to KZG for proto-danksharding` thread on EthResearch:
+
+```text
+The Poseidon hash function was officially introduced in 2019. Since then it has seen considerable attempts at cryptanalysis and optimization. However, it is still very young compared to popular “traditional” hash functions (eg. SHA256 and Keccak), and its general approach of accepting a high level of algebraic structure to minimize constraint count is relatively untested.
+There are layer-2 systems live on the Ethereum network and other systems that already rely on these hashes for their security, and so far they have seen no bugs for this reason. Use of Poseidon in production is still somewhat “brave” compared to decades-old tried-and-tested hash functions, but this risk should be weighed against the risks of proposed alternatives (eg. pairings with trusted setups) and the risks associated with centralization that might come as a result of dependence on powerful provers that can prove SHA256.
+```
```

Comment:
```suggestion
> The Poseidon hash function was officially introduced in 2019. Since then it has seen considerable attempts at cryptanalysis and optimization. However, it is still very young compared to popular “traditional” hash functions (eg. SHA256 and Keccak), and its general approach of accepting a high level of algebraic structure to minimize constraint count is relatively untested.
> There are layer-2 systems live on the Ethereum network and other systems that already rely on these hashes for their security, and so far they have seen no bugs for this reason. Use of Poseidon in production is still somewhat “brave” compared to decades-old tried-and-tested hash functions, but this risk should be weighed against the risks of proposed alternatives (eg. pairings with trusted setups) and the risks associated with centralization that might come as a result of dependence on powerful provers that can prove SHA256.
```
Use a blockquote.